commit 1085e5e17b7403b15c4eff397f3c98c15660f46d
Author: Jacob Walls <jacobtylerwalls@gmail.com>
Date:   Fri May 1 12:20:28 2026 -0400

    Fixed #36300 -- Restored the semantic where RemoteUserMiddleware.header corresponds to request.META under ASGI.
    
    Because these tests always passed both WSGI environ values and HTTP
    headers via `**extra`, this masked a behavior difference between WSGI
    and ASGI.
    
    What should happen: everything should be passed via `headers` but for
    the default REMOTE_USER case on WSGI, which should be passed via
    `**extra`.
    
    Since that was not done, a regression made it into Django 5.2
    (50f89ae850f6b4e35819fe725a08c7e579bfd099) where `.header` no longer
    corresponded to the request.META key under ASGI. To cope, an ASGI user
    would have started(*) sending HTTP headers that match the `.header`
    attribute, which may or may not have been edited to remove the HTTP_
    prefix. (Note: the default `REMOTE_USER` case did not work under ASGI,
    so the change in Django 5.2 had the effect of fixing the default case
    but changing the semantic of the custom case.)
    
    (*): Unless they were getting the sync execution path, which didn't have
    this bug. See the fix in 0f4fff79d33b7cc84822e66bd1fc16caf8222e3a.
    
    Thanks Mykhailo Havelia and Sarah Boyce for reviews.

diff --git a/tests/auth_tests/test_remote_user.py b/tests/auth_tests/test_remote_user.py
index 8c85c65196..69b4b5e10f 100644
--- a/tests/auth_tests/test_remote_user.py
+++ b/tests/auth_tests/test_remote_user.py
@@ -31,8 +31,15 @@ def sync_middleware(get_response):
 class RemoteUserTest(TestCase):
     middleware = "django.contrib.auth.middleware.RemoteUserMiddleware"
     backend = "django.contrib.auth.backends.RemoteUserBackend"
+
+    # ASGI tests always provide this value via `headers`.
+    # WSGI tests provide this value directly into the environ via `**extra`.
+    # When subclassing to provide a custom header, implement a property to
+    # strip the prefix, and always pass via `headers`, since the only case
+    # where it should be passed via the environ is under WSGI, and only when
+    # the value is exactly "REMOTE_USER".
     header = "REMOTE_USER"
-    email_header = "REMOTE_EMAIL"
+    email_header = "HTTP_REMOTE_EMAIL"
 
     # Usernames to be passed in REMOTE_USER for the test_known_user test case.
     known_user = "knownuser"
@@ -77,7 +84,9 @@ class RemoteUserTest(TestCase):
         self.assertTrue(response.context["user"].is_anonymous)
         self.assertEqual(await User.objects.acount(), num_users)
 
-        response = await self.async_client.get("/remote_user/", **{self.header: ""})
+        response = await self.async_client.get(
+            "/remote_user/", headers={self.header: ""}
+        )
         self.assertTrue(response.context["user"].is_anonymous)
         self.assertEqual(await User.objects.acount(), num_users)
 
@@ -154,7 +163,7 @@ class RemoteUserTest(TestCase):
         """See test_unknown_user."""
         num_users = await User.objects.acount()
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: "newuser"}
+            "/remote_user/", headers={self.header: "newuser"}
         )
         self.assertEqual(response.context["user"].username, "newuser")
         self.assertEqual(await User.objects.acount(), num_users + 1)
@@ -162,7 +171,7 @@ class RemoteUserTest(TestCase):
 
         # Another request with same user should not create any new users.
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: "newuser"}
+            "/remote_user/", headers={self.header: "newuser"}
         )
         self.assertEqual(await User.objects.acount(), num_users + 1)
 
@@ -188,14 +197,14 @@ class RemoteUserTest(TestCase):
         await User.objects.acreate(username="knownuser2")
         num_users = await User.objects.acount()
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(response.context["user"].username, "knownuser")
         self.assertEqual(await User.objects.acount(), num_users)
         # A different user passed in the headers causes the new user
         # to be logged in.
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user2}
+            "/remote_user/", headers={self.header: self.known_user2}
         )
         self.assertEqual(response.context["user"].username, "knownuser2")
         self.assertEqual(await User.objects.acount(), num_users)
@@ -233,7 +242,7 @@ class RemoteUserTest(TestCase):
         await user.asave()
 
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertNotEqual(default_login, response.context["user"].last_login)
 
@@ -241,7 +250,7 @@ class RemoteUserTest(TestCase):
         user.last_login = default_login
         await user.asave()
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(default_login, response.context["user"].last_login)
 
@@ -271,7 +280,7 @@ class RemoteUserTest(TestCase):
         await User.objects.acreate(username="knownuser")
         # Known user authenticates
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(response.context["user"].username, "knownuser")
         # During the session, the REMOTE_USER header disappears. Should trigger
@@ -307,12 +316,12 @@ class RemoteUserTest(TestCase):
         await User.objects.acreate(username="knownuser")
         # Known user authenticates
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(response.context["user"].username, "knownuser")
         # During the session, the REMOTE_USER changes to a different user.
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: "newnewuser"}
+            "/remote_user/", headers={self.header: "newnewuser"}
         )
         # The current user is not the prior remote_user.
         # In backends that create a new user, username is "newnewuser"
@@ -327,7 +336,7 @@ class RemoteUserTest(TestCase):
     async def test_inactive_user_async(self):
         await User.objects.acreate(username="knownuser", is_active=False)
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: "knownuser"}
+            "/remote_user/", headers={self.header: "knownuser"}
         )
         self.assertTrue(response.context["user"].is_anonymous)
 
@@ -355,7 +364,7 @@ class RemoteUserNoCreateTest(RemoteUserTest):
     async def test_unknown_user_async(self):
         num_users = await User.objects.acount()
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: "newuser"}
+            "/remote_user/", headers={self.header: "newuser"}
         )
         self.assertTrue(response.context["user"].is_anonymous)
         self.assertEqual(await User.objects.acount(), num_users)
@@ -374,7 +383,7 @@ class AllowAllUsersRemoteUserBackendTest(RemoteUserTest):
     async def test_inactive_user_async(self):
         user = await User.objects.acreate(username="knownuser", is_active=False)
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(response.context["user"].username, user.username)
 
@@ -402,7 +411,7 @@ class CustomRemoteUserBackend(RemoteUserBackend):
         return user
 
     async def aconfigure_user(self, request, user, created=True):
-        user.email = request.META.get("HTTP_" + RemoteUserTest.email_header, "")
+        user.email = request.META.get(RemoteUserTest.email_header, "")
         if not created:
             user.last_name = user.username
         await user.asave()
@@ -468,9 +477,9 @@ class RemoteUserCustomTest(RemoteUserTest):
         num_users = await User.objects.acount()
         response = await self.async_client.get(
             "/remote_user/",
-            **{
+            headers={
                 self.header: "newuser",
-                self.email_header: "user@example.com",
+                self.email_header.removeprefix("HTTP_"): "user@example.com",
             },
         )
         self.assertEqual(response.context["user"].username, "newuser")
@@ -503,19 +512,51 @@ class CustomHeaderMiddleware(RemoteUserMiddleware):
     header = "HTTP_AUTHUSER"
 
 
-class CustomHeaderRemoteUserTest(RemoteUserTest):
+class CustomHeaderMixin:
+    middleware = "auth_tests.test_remote_user.CustomHeaderMiddleware"
+    auth_header = CustomHeaderMiddleware.header
+
+    @property
+    def header(self):
+        """Return the unprefixed header the server or proxy should provide."""
+        return self.auth_header.removeprefix("HTTP_")
+
+    def setUp(self):
+        """Force any **kwargs to be passed via `headers`."""
+        super().setUp()
+        original_get = self.client.get
+        original_async_get = self.async_client.get
+
+        def get(self, *args, headers={}, **kwargs):
+            headers = {**headers}
+            headers.update(kwargs)
+            return original_get(self, *args, headers=headers)
+
+        async def aget(self, *args, headers={}, **kwargs):
+            headers = {**headers}
+            headers.update(kwargs)
+            return await original_async_get(self, *args, headers=headers)
+
+        self.client.get = get
+        self.async_client.get = aget
+        self.addCleanup(setattr, self.client, "get", original_get)
+        self.addCleanup(setattr, self.async_client, "get", original_async_get)
+
+
+class CustomHeaderRemoteUserTest(CustomHeaderMixin, RemoteUserTest):
     """
     Tests a custom RemoteUserMiddleware subclass with custom HTTP auth user
     header.
     """
 
-    middleware = "auth_tests.test_remote_user.CustomHeaderMiddleware"
-    header = "HTTP_AUTHUSER"
 
-
-class CustomHeaderASGISyncPathRemoteUserTest(ASGISyncPathRemoteUserTest):
-    middleware = "auth_tests.test_remote_user.CustomHeaderMiddleware"
-    header = "HTTP_AUTHUSER"
+class CustomHeaderASGISyncPathRemoteUserTest(
+    CustomHeaderMixin, ASGISyncPathRemoteUserTest
+):
+    """
+    Tests sync execution under ASGI of a custom RemoteUserMiddleware subclass
+    with custom HTTP auth user header.
+    """
 
 
 class PersistentRemoteUserTest(RemoteUserTest):
@@ -546,7 +587,7 @@ class PersistentRemoteUserTest(RemoteUserTest):
         await User.objects.acreate(username="knownuser")
         # Known user authenticates
         response = await self.async_client.get(
-            "/remote_user/", **{self.header: self.known_user}
+            "/remote_user/", headers={self.header: self.known_user}
         )
         self.assertEqual(response.context["user"].username, "knownuser")
         # Should stay logged in if the REMOTE_USER header disappears.
